### ISSUE TITLE
prevent 🙈 when deploying client side changes

### ISSUE THREAD

**Issue by homeyer** [waffleio/waffle.io#3273]
Problem: When we do a rolling deploy of the app k8s, requests for the new client bundle is guaranteed to be served a new pod.

AC:
- [ ] app does not do rolling deployments. When an app deploy happens, requests should immediately switch from an old deployment to the new. 

Could you cordon new pods as you create them, and then uncordon new and cordon old at the same time?

Should we maintain a blue/green and somehow switch back and forth during deploys? I'd rather not introduce unnecessary resources though for cost.

**Comment by jtslear** [comment]
In general, I think the safest approach to this would be to allow those assets to continue to be available.  The best way I can think of, would be to revert, at least a portion, of the work done removing out aws infra that allowed this problem to never show up.

Other options:
## CanRolli
* This would require a bit of playing around to get right, but the goal would be to muck around with the timings for the Rolling Deploy mechanism such that things are flipped over at the right time
* Adjust our options for the readiness probe, setting `initialDelay` to something that takes into account the time required to do the following: pull the images + start up + be ready to accept connections
* Set the `strategy.rollingupdate.maxSurge` to 100%.  
* With these two options set, kubernetes will perform a deploy like so:
  * version 1 will already be running with 8 pods
  * version 2 will ask to be deployed
  * version 1 will spin down 25% (down to 6 pods); version 2 will start creating 8 pods
  * version 1 will be running 6 pods, version 2 will be running 8 pods
    * waiting on readiness probe initialDelay timer to expire
    * all traffic will still be on version 1 since version 2 is not considered ready
  * Assuming all is well with version 2, the readiness probe initialDelay will expire
  * healthchecks will pass, version 1 will all be terminated at the same time
    * at this point, we'll probably see some failed replies from the version 1 pods
    * all new traffic is sent to version2 since any remaining version 1 pods are in state terminating
### Pros
* easy

### Cons
* there still exists the potential a request may go to version 1, but we've limited that ability greatly

## Consommé
* This would require a lot of interaction with kubernetes and it's api.  This is probably outside the boundaries of a simple bash script deploy.  Complexity level high, but in testing, it works very well (all testing done manually)
* The goal here would be to simply point the service to the right set of pods at the right time
  * version 1 would already be running with 8 pods
    * the service declaratation is currently told to use version 1; so all traffic goes here
  * version 2 creates a brand new deployment and spins up 8 pods
    * the service declaration should have no clue what this is, so no traffic actually goes to it
  * after we know version 2 works, we flip a switch in the service declaration 
  * all traffic will immediately be going to version 2
  * we then tear down version 1
### Pros
* it sounds awesome actually
* there should not be ANY possibility of version 1 being used at all

### Cons
* more difficult to implement
* I'm not sure what happens to requests that went to version 1.  Do they get completed? or left dangling?

## Spaghetti
* The goal here is to actually leverage the app and browser a bit
* Since our problem is the actually resource `assets.json` telling the browser what assets to pull, what if we cache it?
* Cloudflare won't, so we'll have to tell the browser to do so
* If we cache for the amount of time it takes for a deploy, we can minimize the impact of an outage

### Pros
* super duper easy
* if someone is requesting an old resource, cloudflare should toss them the cached version

### Cons
* we are kinda relying on time
* cache busting won't help us with the asset.json file...

**Comment by homeyer** [comment]
I emailed cloudflare support to verify they actually support HTTP/2 Server Push, because afaik they do not.

my email -----

Hi!

I'm not able to get cloudflare's HTTP/2 Server Push support working on our site.

I went searching for an answer, and stumbled on your blog post: https://blog.cloudflare.com/http-2-server-push-with-multiple-assets-per-link-header/. The post itself is configured to send assets via push.

The page (the blog post) is sending a link header with assets, but cloudflare isn't adding the expected cf-h2-pushed header, and Chrome's network tab doesn't show any assets being pushed.

I verified that my browser is loading it as h2 via https://blog.cloudflare.com/cdn-cgi/trace. I'm on Chrome 61.

Is HTTP/2 push still supported by Cloudflare?
